ESQL: Views REST API

craigtaverner opened the pull request:
This is extracted from the Views Prototype at https://github.com/elastic/elasticsearch/pull/134995 and is originally based on the "ESQL: View" prototype by Nik but with the following changes:
* No ES|QL language integration, only REST API for adding/removing and managing Views in cluster state
* ViewMetadata now inside ProjectMetadata
* NodeFeature to ensure metadata only exists in clusters fully upgraded
* Wide CRUD support (add, update, delete, list)
* View name validation
* View count, length and depth restrictions

The REST API looks like:

```
PUT _query/view/critical_service_errors_v {"query":"FROM ..."}
DELETE _query/view/critical_service_errors_v
GET _query/view/critical_service_errors_v
GET _query/view
```

Checklist:

 - [x] Extract REST API from views prototype
 - [x] Remove `list` interface and generalise GET command to cover list use cases (more in-line with other APIs)
 - [x] Verify that snapshot-only really blocks registration of REST API
 - [x] Consider switching to use feature flags (instead of snapshot-only)
 - [x] Consider dropping one of NodeFeature or TransportVersion since they cover the same requirement
 - [x] Implement batched tasks, probably using `SequentialAckingBatchedTaskExecutor `
 - [x] Stop using `ProjectResolver` and instead use the `ProjectID` as is done in Enrich REST API
 - [x] Change ViewMetadata from a Map of name to View to instead be an array of View (which should contain the name)

craigtaverner committed:
ESQL: Views REST API

This is extracted from the Views Prototype at https://github.com/elastic/elasticsearch/pull/134995
and is originally based on the "ESQL: View" prototype by Nik but with the following changes:
* No ES|QL language integration, only REST API for adding/removing and managing Views in cluster state
* ViewMetadata now inside ProjectMetadata
* NodeFeature to ensure metadata only exists in clusters fully upgraded
* Wide CRUD support (add, update, delete, list)
* View name validation
* View count, length and depth restrictions

Co-authored-by: Nik Everett <nik9000@gmail.com>

craigtaverner committed:
Update docs/changelog/137818.yaml

craigtaverner committed:
Refine changelog

craigtaverner committed:
Expand get-views to cover use cases for listing views

craigtaverner committed:
Fixed failing tests after changing behaviour of get to throw exception on missing view

craigtaverner committed:
Fixed failing yaml-tests after changing behaviour of get to return older list structures

craigtaverner committed:
Remove list_views API (covered by get_view now)

nielsbauman commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/InMemoryViewService.java:
```diff
@@ -0,0 +1,48 @@
+ * Simple implementation of {@link ClusterViewService} that keeps the views in memory.
+ * This is useful for testing.
+ */
+public class InMemoryViewService extends ViewService {
```
I'm probably missing something, but why do we need this in-memory version exactly?

And, follow-up question: wouldn't it make more sense to put that in a test package somewhere, rather than having it in production code, if it's only useful for testing?

craigtaverner replied:
It is only moderately useful here, but very useful in the actual ES|QL views were we have semi-unit tests that run the full ES|QL stack, but outside of Elasticsearch, and then this becomes the main mechanism for enabling that. I could indeed move it.

nielsbauman reviewed (commented):
Thanks for working on this, @craigtaverner! I scanned most parts of the code briefly and left a few comments, some of which are not a big deal.

craigtaverner commented on x-pack/plugin/security/qa/operator-privileges-tests/src/javaRestTest/java/org/elasticsearch/xpack/security/operator/Constants.java:
```diff
@@ -174,6 +174,10 @@ public class Constants {
+        "cluster:admin/xpack/esql/view/put",
+        "cluster:admin/xpack/esql/view/delete",
+        "cluster:admin/xpack/esql/view/get",
+        "cluster:internal/xpack/inference/clear_inference_endpoint_cache",
```
Ah, yes that was removed in main, but my conflict resolution failed to remove it in my branch. I'll fix that.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/ViewMetadata.java:
```diff
@@ -0,0 +1,121 @@
+                }
+            }
+            return patterns;
+        }, VIEWS);
```
This is copied verbatim from EnrichMetadata, which has a very similar structure with the constructor taking a Map<String, EnrichPolicy>, and here we take Map<String, View>. The difference is handled on line 56 which decodes the View using View.fromXContent. You suggest that I skip the map, which I'm not sure makes sense to me.

craigtaverner committed:
Simply fixes based on code review

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into views_rest_crud

craigtaverner committed:
Fix transport version numbers

nielsbauman commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/ClusterViewService.java:
```diff
@@ -0,0 +1,96 @@
+    @SuppressForbidden(reason = "legacy usage of unbatched task") // TODO add support for batching here
+    private void submitUnbatchedTask(@SuppressWarnings("SameParameterValue") String source, ClusterStateUpdateTask task) {
+        clusterService.submitUnbatchedStateUpdateTask(source, task);
+    }
```
Yeah that code is definitely not recent either, hence the `TODO` in the code :)

craigtaverner committed:
Update REST API to use ProjectID where possible

craigtaverner committed:
Fixed things that did not get backported correctly

craigtaverner committed:
Fixed named writable

craigtaverner committed:
Missing project ID

quackaplop commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/EsqlFeatures.java:
```diff
@@ -34,9 +34,15 @@ public class EsqlFeatures implements FeatureSpecification {
     private Set<NodeFeature> snapshotBuildFeatures() {
         assert Build.current().isSnapshot() : Build.current();
-        return Set.of(METRICS_SYNTAX);
+        return Set.of(METRICS_SYNTAX, ESQL_VIEWS);
```
I forget, did we decide we need this as NodeFeature or not?

craigtaverner committed:
Register NamedDiff

craigtaverner committed:
Make views tests not fail in release mode

craigtaverner committed:
Disable for release build and have tests that assert this

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into views_rest_crud

craigtaverner committed:
Update transport versions

craigtaverner committed:
Switch to using FeatureFlag for enabling/disabling the API

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into views_rest_crud

craigtaverner committed:
Update transport versions

craigtaverner committed:
Merge branch 'main' into views_rest_crud

craigtaverner committed:
Do not use cluster_features for yaml tests

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into views_rest_crud

craigtaverner committed:
Remove NodeFeature because the FeatureFlag conflicts with it

Since the feature flag turns the feature off in snapshots, the NodeFeature is missing in release builds, which makes it unusable in YAML tests, unlike capabilities. While we can access the node feature via the capabilities API, and thereby fix the YAML tests, this removes the remaining reason for having the NodeFeature at all. We already have a TransportVersion check, which covers the other point of a NodeFeature.

craigtaverner committed:
Merge branch 'main' into views_rest_crud

craigtaverner committed:
Merge branch 'main' into views_rest_crud

dakrone commented on docs/changelog/137818.yaml:
```diff
@@ -0,0 +1,5 @@
+summary: ES|QL Views REST API
+area: "ES|QL"
+type: feature
+issues: []
```
For now this PR should be marked as `>non-issue` so that no changelog is generated. That way it does not show up in the public changelog until the feature flag has been removed (that one will have `>feature` and `>release-highlight` on it).

dakrone commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/GetViewAction.java:
```diff
@@ -0,0 +1,126 @@
+        }
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
```
Feels like this should throw an exception if we try to serialize this, since it doesn't actually implement the serialization.

dakrone commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/TransportGetViewAction.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.List;
+import java.util.TreeMap;
+
+public class TransportGetViewAction extends HandledTransportAction<GetViewAction.Request, GetViewAction.Response> {
```
I believe this would be better if it extended `TransportLocalProjectMetadataAction` instead. This would mean that your `localClusterStateOperation` method will get the appropriate `ProjectState` automatically, and there will be no need to use the `viewService.getProjectId()` method.

dakrone commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/TransportGetViewAction.java:
```diff
@@ -0,0 +1,59 @@
+
+    @Override
+    protected void doExecute(Task task, GetViewAction.Request request, ActionListener<GetViewAction.Response> listener) {
+        ProjectId projectId = viewService.getProjectId();
```
I mentioned above, but seeing that we are retrieving a project id here from the view service rather than injecting the project resolver directly into the transport action seems like a smell to me. We can avoid it by using the `TransportLocalProjectMetadataAction` I mentioned above.

craigtaverner replied:
Done as part of the above change

dakrone commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/TransportGetViewAction.java:
```diff
@@ -0,0 +1,59 @@
+            }
+        }
+        if (missing.isEmpty() == false) {
+            listener.onFailure(new IllegalArgumentException("Views do not exist: " + String.join(", ", missing)));
```
This should use `ResourceNotFoundException` instead of `IllegalArgumentException` here to ensure the correct HTTP response code.

dakrone reviewed (dismissed):
I've added some comments for this. One thing that I don't see at all is the overlap checks to ensure that we do not create a view that has the same name as an index, alias, data stream, or data stream alias.

craigtaverner committed:
Move from unbatch to batched action processing using SequentialAckingBatchedTaskExecutor

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/ViewService.java:
```diff
@@ -0,0 +1,243 @@
+        }
+        new EsqlParser().createStatement(view.query(), new QueryParams(), new PlanTelemetry(functionRegistry));
+        // TODO should we validate this in the transport action and make it async? like plan like a query
+        // TODO postgresql does.
```
This comment comes from Nik's original prototype, and I'd planned to revisit it and think it through some more. I agree that it should not be asynchronous. View CRUD does not need to be especially responsive.

craigtaverner committed:
Round of fixes based on code review

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/ViewService.java:
```diff
@@ -0,0 +1,243 @@
+import java.util.Set;
+import java.util.function.Function;
+
+public abstract class ViewService {
```
The original reason for doing this was that in ES|QL we have a large suite of csv-spec tests that are run both as full integration tests and as unit tests with only the ES|QL code activated. For the unit tests version, we need a way to mock the view service, and the InMemoryViewService was that solution. We want it to share as much code with the production version, but not have access to cluster state. It only keeps the map of names to queries. So this abstraction is very useful for ES|QL tests. In this PR it only allows for the InMemoryViewServiceTests, which is still useful, but less so.

craigtaverner committed:
Merge branch 'main' into views_rest_crud

craigtaverner committed:
Delete docs/changelog/137818.yaml

craigtaverner committed:
Fixed faioling tests

craigtaverner committed:
Merge branch 'main' into views_rest_crud

craigtaverner committed:
Simplify metadata parsing

The original code came from a modification of EnrichPolicyMetadata, which was unnecessarily complex for this case.

craigtaverner committed:
Fix comment and add one more syntax validation test

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into views_rest_crud

craigtaverner committed:
Update transport version

craigtaverner committed:
Simplify view service

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into views_rest_crud

craigtaverner committed:
Update transport version

craigtaverner committed:
Merge branch 'main' into views_rest_crud

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into views_rest_crud

craigtaverner committed:
Fixes after merging main

nielsbauman commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/ClusterViewService.java:
```diff
@@ -0,0 +1,119 @@
+
+        @Override
+        public ClusterState execute(ClusterState currentState) {
+            var project = getProjectMetadata(projectId);
```
Using `getProjectMetadata(projectId)` here is wrong, as it gets the `ClusterState` from the `ClusterService` instead of using the one in this method. That means it will use an outdated cluster state to compute the updated cluster state. This is another reason why I'm not a huge fan of having a method like `getProjectMetadata` in this class, as it's easier to introduce bugs.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into views_rest_crud

craigtaverner committed:
Update transport version

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/ClusterViewService.java:
```diff
@@ -0,0 +1,119 @@
+        public ClusterState execute(ClusterState currentState) {
+            var project = getProjectMetadata(projectId);
+            var views = project.custom(ViewMetadata.TYPE, ViewMetadata.EMPTY);
+            Map<String, View> policies = updateFunction.apply(views);
```
This abstraction actually dates back to Nik's original prototype. I personally like it, but can change if others feel strongly about it.

craigtaverner committed:
Fixed getting ProjectMetadata from cluster service

Get it from the current ClusterState instead

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into views_rest_crud

craigtaverner committed:
Updated transport version

craigtaverner committed:
Updated ViewMetadata with view name inside View, and structured as list

Changes the structure to look like:
{
  "views": [
    {
      "name": "view1",
      "query": "FROM …"
    },
    {
      "name": "otherview",
      "query": "FROM …"
    }
  ]
}

craigtaverner committed:
Updated GET views to use new array format

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into views_rest_crud

craigtaverner committed:
Update transport version

craigtaverner committed:
Merge branch 'main' into views_rest_crud

craigtaverner committed:
Merge branch 'main' into views_rest_crud

craigtaverner committed:
Merge branch 'main' into views_rest_crud

craigtaverner committed:
Moved View and ViewMetadata to server

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into views_rest_crud

craigtaverner committed:
Update transport version

craigtaverner committed:
Merge branch 'views_rest_crud' of github.com:craigtaverner/elasticsearch into views_rest_crud

craigtaverner committed:
Move ViewTests and ViewMetadataTests to server

craigtaverner committed:
Merge branch 'main' into views_rest_crud

craigtaverner committed:
Merge branch 'main' into views_rest_crud

dakrone committed:
Start making things use maps

dakrone committed:
Collapse ViewService into a single service (removing ClusterViewService)

dakrone committed:
Refactor settings infrastructure to register dynamic settings in plugin

dakrone committed:
Embrace asynchronicity, rewrite InMemoryViewService and tests

dakrone committed:
AwaitsFix something for now

dakrone committed:
Fix ViewMetadata serialization

dakrone committed:
Fix View serialization in get-view response

dakrone committed:
Remove ViewIterator

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into views_rest_crud

craigtaverner committed:
After merging main, updated transport version and fixed tests

craigtaverner committed:
Only create ViewService component if feature is enabled

dakrone committed:
Get the yaml tests to actually run

dakrone committed:
Fix crud yaml test

dakrone committed:
Remove unnecessary tests, move validation into yaml tests

dakrone committed:
Remove view functionality test, as this is not yet implemented

dakrone committed:
Merge remote-tracking branch 'origin/main' into views_rest_crud

dakrone committed:
Stricter validation

dakrone committed:
Cleanups

dakrone committed:
harden test

dakrone committed:
Generate a valid view name (lowercase) instead of random alpha

dakrone committed:
Stricter validation at REST layer

dakrone committed:
Fix json spec for put view visability

dakrone committed:
no need to wrap listener and fake always being acknowledged

dakrone committed:
Fix comment

dakrone committed:
Fix missing inference settings after merging main

dakrone committed:
Merge remote-tracking branch 'origin/main' into views_rest_crud

dakrone committed:
Add test features to fix tests

nielsbauman commented on server/src/test/java/org/elasticsearch/cluster/metadata/ViewMetadataTests.java:
```diff
@@ -0,0 +1,61 @@
+    @Override
+    protected ViewMetadata createXContextTestInstance(XContentType xContentType) {
+        return randomViewMetadata();
+    }
```
Nit: no need to override this method; that's already the default: https://github.com/elastic/elasticsearch/blob/c6f47cc2184edbbb4d6869beacc8459a1624838b/test/framework/src/main/java/org/elasticsearch/test/AbstractChunkedSerializingTestCase.java#L37-L40
```suggestion
```

nielsbauman commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/TransportGetViewAction.java:
```diff
@@ -0,0 +1,88 @@
+            } else {
+                views.add(view);
+            }
+        }
```
May I suggest something like:
```suggestion
        Collection<View> views = new ArrayList<>();
        List<String> missing = new ArrayList<>();
        Collection<String> names = request.names();
        if (names.isEmpty()) {
            views = viewService.getMetadata(project).views().values();
        } else {
	        for (String name : names) {
	            View view = viewService.get(projectId, name);
	            if (view == null) {
	                missing.add(name);
	            } else {
	                views.add(view);
	            }
	        }
	    }
```
It feels a bit wasteful to loop over all the views again if we're just retrieving all views. This is just a suggestion, I won't block the PR on this if either of you prefers the current code.

nielsbauman commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/ViewService.java:
```diff
@@ -0,0 +1,321 @@
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                final ProjectMetadata project = currentState.metadata().getProject(projectId);
+                final ViewMetadata viewMetadata = project.custom(ViewMetadata.TYPE, ViewMetadata.EMPTY);
```
Might as well make use of the helpers we add :D
```suggestion
                final ViewMetadata viewMetadata = getMetadata(project);
```

nielsbauman commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/ViewService.java:
```diff
@@ -0,0 +1,321 @@
+        MetadataCreateIndexService.validateIndexOrAliasName(
+            name,
+            (viewName, error) -> new IllegalArgumentException("invalid view name [" + name + "], " + error)
+        );
```
I don't think there's a lot of value in this check; is there? We just use the name for some map lookups and, fortunately, there's no such thing as Java-Map-injections. We do output the name in the failure message below; is that why you have this added check?

It feels a bit random here to have this check here, but not do the proper name validation we do in `PutViewAction` (i.e. the lowercase check).

dakrone replied:
Yes I agree, I do not like having the duplicate validation (and it's already in the REST side anyway).

nielsbauman reviewed (commented):
Hi both, I reviewed all the production code for now. I'll have a look at the test code later, but I figured I'd post my comments thus far - it's mostly minor stuff!

dakrone committed:
Remove unused constructor

dakrone committed:
Widen ViewMetadataTests coverage to empty views

dakrone committed:
Remove needlessly overridden method

dakrone committed:
Remove GetViewAction.Request(StreamInput)

dakrone committed:
Avoid double loop when retrieving names

dakrone committed:
Use metadata helper to retrieve metadata in CSU

dakrone committed:
Add no-op check to putView

dakrone committed:
Remove re-validating in the service (already validated at REST layer) and test

dakrone committed:
Remove duplicated validation

dakrone committed:
Merge remote-tracking branch 'origin/main' into views_rest_crud

nielsbauman commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/view/ViewRestTests.java:
```diff
@@ -0,0 +1,41 @@
+
+import static org.elasticsearch.xpack.esql.plugin.EsqlFeatures.ESQL_VIEWS_FEATURE_FLAG;
+
+public class ViewRestTests extends AbstractViewTestCase {
```
Are all the changes in `‎x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/view` only for this test class? Or are you planning on adding more tests in the future? The amount of plumbing/infra we do for this test is excessive and unnecessary; that's why I'm asking.

nielsbauman reviewed (commented):
Left a few more minor comments, having reviewed everything now. Almost good to go from my side!

dakrone committed:
Use helper to resolve metadata

dakrone committed:
Remove extra cluster feature requirements in yaml tests

dakrone committed:
Assert view length in yaml tests

dakrone committed:
Merge remote-tracking branch 'origin/main' into views_rest_crud

nielsbauman committed:
Use Capabilities API instead of cluster feature

nielsbauman reviewed (approved):
LGTM, thanks a lot for all the iterations both @craigtaverner and @dakrone!

dakrone committed:
Merge remote-tracking branch 'origin/main' into views_rest_crud

dakrone committed:
Merge remote-tracking branch 'origin/main' into views_rest_crud

dakrone committed:
Merge remote-tracking branch 'origin/main' into views_rest_crud
